### PR TITLE
Multiple histogram

### PR DESCRIPTION
**Pull request by menghanl**
- Simply histogram, remove counter and tracker.

- Create one histogram for each goroutine. Merge histograms when reporting stats.

- Small changes in function getStats(), add bool argument and remove function reset() in both client and server.

### CODE REVIEW

File: benchmark/stats/histogram.go [hunk 1]
```diff
@@ -106,11 +47,12 @@ type HistogramOptions struct {
 	MinValue int64
 }
 
-// bucketInternal is the internal representation of a bucket, which includes a
-// rate counter.
-type bucketInternal struct {
-	lowBound float64
-	count    *Counter
+// HistogramBucket is one histogram bucket.
```
Comment (iamqizhao) [inline benchmark/stats/histogram.go]: HistogramBucket represents a bucket in the histogram.

File: benchmark/stats/histogram.go [hunk 2]
```diff
@@ -4,14 +4,18 @@ import (
...
 	"strconv"
 	"strings"
-	"time"
 )
 
-// HistogramValue is the value of Histogram objects.
-type HistogramValue struct {
+// Histogram accumulates values in the form of a histogram with
+// exponentially increased bucket sizes.
+// The first bucket (with index 0) is [0, n) where n = baseBucketSize.
```
Comment (iamqizhao) [inline benchmark/stats/histogram.go]: [HistogramOptions.MinValue, HistogramOptions.MinValue + n) and move this to the comments of HistogramOptions.

File: benchmark/stats/histogram.go [hunk 3]
```diff
@@ -123,36 +66,75 @@ func NewHistogram(opts HistogramOptions) *Histogram {
...
-		h.buckets[i].lowBound = float64(opts.MinValue) + delta
-		h.buckets[i].count = newCounter()
+		h.Buckets[i].LowBound = float64(opts.MinValue) + delta
 		delta = delta * m
 	}
 	return &h
 }
 
+// Print writes textual output of the histogram values.
+func (h Histogram) Print(w io.Writer) {
```
Comment (iamqizhao) [inline benchmark/stats/histogram.go]: u need to make the receiver consistent. You need to use pointer receiver for consistency.

File: benchmark/stats/histogram.go [hunk 4]
```diff
@@ -273,8 +171,28 @@ func (h *Histogram) findBucket(value int64) (int, error) {
...
 		b = int((math.Log(delta)-h.logBaseBucketSize)*h.oneOverLogOnePlusGrowthFactor + 1)
 	}
-	if b >= len(h.buckets) {
+	if b >= len(h.Buckets) {
 		return 0, fmt.Errorf("no bucket for value: %d", value)
 	}
 	return b, nil
 }
+
+// Merge takes another histogram, and merges its content into the receiver.
```
Comment (iamqizhao) [inline benchmark/stats/histogram.go]: Merge merges h2 into h.

File: benchmark/worker/benchmark_client.go [hunk 5]
```diff
@@ -54,11 +54,12 @@ var (
...
-	stop          chan bool
-	mu            sync.RWMutex
-	lastResetTime time.Time
-	histogram     *stats.Histogram
+	closeConns       func()
+	stop             chan bool
+	lastResetTime    time.Time
+	histogramOptions stats.HistogramOptions
+	mutexes          []*sync.Mutex
+	histograms       []*stats.Histogram
```
Comment (iamqizhao) [inline benchmark/worker/benchmark_client.go]: can you do 

hs   []*lockingHistogram

where

type lockHistogram {
  mu  sync.Mutex
  h     *stats.Histogram
}

File: benchmark/worker/benchmark_client.go [hunk 6]
```diff
@@ -199,13 +200,17 @@ func startBenchmarkClient(config *testpb.ClientConfig) (*benchmarkClient, error)
...
-		histogram: stats.NewHistogram(stats.HistogramOptions{
+		histogramOptions: stats.HistogramOptions{
 			NumBuckets:     int(math.Log(config.HistogramParams.MaxPossible)/math.Log(1+config.HistogramParams.Resolution)) + 1,
 			GrowthFactor:   config.HistogramParams.Resolution,
 			BaseBucketSize: (1 + config.HistogramParams.Resolution),
 			MinValue:       0,
-		}),
+		},
+		mutexes:    make([]*sync.Mutex, rpcCountPerConn*len(conns)),
+		histograms: make([]*stats.Histogram, rpcCountPerConn*len(conns)),
```
Comment (iamqizhao) [inline benchmark/worker/benchmark_client.go]: make([]lockingHistogram, rpcCountPerConn*len(conns), rpcCountPerConn*len(conns))

specify the cap to eliminate unnecessary memory footprint.

**Comment by menghanl** [comment]
Thanks for the review. All fixed. PTAL.

File: benchmark/worker/benchmark_client.go [hunk 7]
```diff
@@ -294,9 +309,9 @@ func (bc *benchmarkClient) doCloseLoopStreaming(conns []*grpc.ClientConn, rpcCou
 							return
 						}
 						elapse := time.Since(start)
-						bc.mu.Lock()
-						bc.histogram.Add(int64(elapse))
-						bc.mu.Unlock()
+						bc.lockingHistograms[idx].mu.Lock()
```
Comment (iamqizhao) [inline benchmark/worker/benchmark_client.go]: make these 3 lines as a member function lockingHistogram.

**Pull request merged**